### PR TITLE
[Fixes #6598] - Updates Write-Output

### LINKED ISSUE #6598: Emphasize the information that Write-Output is normally not needed, clarify when it is useful, use a better example
> This cmdlet is typically used in scripts to display strings and other objects on the console. However, because the default behavior is to display the objects at the end of a pipeline, it is generally not necessary to use the cmdlet. For instance, `Get-Process | Write-Output` is equivalent to `Get-Process`.

All of this is correct, but I suggest:

* highlighting the paragraph with `[!NOTE]` formatting

* using a better example for the benefit of users coming from traditional shells such as `cmd.exe` and `bash`: 
  * Point out that the built-in alias of `Write-Output` is `echo`, and that commands such as 
`echo "Home directory: $HOME"` can be written as simply `"Home directory: $HOME"` - no need for `echo` / `Write-Output`.

* adding the one - rare - use case where `Write-Output` _is_ useful: to output a collection _as a whole_ to the pipeline; e.g., `Write-Output -NoEnumerate 1, 2, 3 | ForEach-Object { "Array: $_ " }` (the more efficient, but more obscure, expression-mode alternative is `, (1, 2, 3)  | ForEach-Object { "Array: $_ " }`

`

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 58ae9a63-c445-5453-7e8c-93121b8fcea8
* Version Independent ID: f0d97751-59f2-7dcb-5c9c-11fa9e557e1b
* Content: [Write-Output (Microsoft.PowerShell.Utility) - PowerShell](https://docs.microsoft.com/en-us/powershell/module/Microsoft.PowerShell.Utility/write-output?view=powershell-7)
* Content Source: [reference/7.0/Microsoft.PowerShell.Utility/Write-Output.md](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/7.0/Microsoft.PowerShell.Utility/Write-Output.md)
* Product: **powershell**
* Technology: **powershell-cmdlets**
* GitHub Login: @JamesWTruher
* Microsoft Alias: **jimtru**

### PR DESCRIPTION
**Pull request by chasewilson**
# PR Summary

Updates Write-Output for some clarity on usage. 

## PR Context

Fixes #6598 
Fixes [AB#1769024](https://dev.azure.com/mseng/677da0fb-b067-4f77-b89b-f32c12bb8617/_workitems/edit/1769024)

Select the area of the Table of Contents containing the documents being changed.

**Conceptual content**
- [ ] Overview and Install
- [ ] Learning PowerShell
  - [ ] PowerShell 101
  - [ ] Deep dives
  - [ ] Remoting
- [ ] Release notes (What's New)
- [ ] Windows PowerShell
  - WMF, ISE, release notes, etc.
- [ ] DSC articles
- [ ] Community resources
- [ ] Sample scripts
- [ ] Gallery articles
- [ ] Scripting and development
  - [ ] Legacy SDK

**Cmdlet reference & about_ topics**
- [x] Version 7.1 preview content
- [x] Version 7.0 content
- [x] Version 6 content
- [x] Version 5.1 content

## PR Checklist

- [x] I have read the [contributors guide][contrib] and followed the style and process guidelines
- [x] PR has a meaningful title
- [x] PR is targeted at the _staging_ branch
- [x] All relevant versions updated
- [x] Includes content related to issues and PRs - see [Closing issues using keywords][key].
- [x] This PR is ready to merge and is not **Work in Progress**
  - If the PR is work in progress, please add the prefix `WIP:` or `[WIP]` to the beginning of the
    title and remove the prefix when the PR is ready.

[contrib]: https://docs.microsoft.com/powershell/scripting/community/contributing/overview
[key]: https://help.github.com/en/articles/closing-issues-using-keywords

### CODE REVIEW

**Comment by chasewilson** [comment]
@mklement0 Feel free to provide feedback here

File: reference/5.1/Microsoft.PowerShell.Utility/Write-Output.md [hunk 1]
```diff
@@ -28,10 +28,12 @@ If the command is the last command in the pipeline, the object is displayed in t
...
-This cmdlet is typically used in scripts to display strings and other objects on the console.
-However, because the default behavior is to display the objects at the end of a pipeline, it is
-generally not necessary to use the cmdlet. For instance, `Get-Process | Write-Output` is equivalent
-to `Get-Process`.
+> [!NOTE] This cmdlet is typically used in scripts to display strings and other objects on the
+> console. One of the built-in aliases for `Write-Output` is `echo` and similar to other shells that
+> use `echo`, the default behavior is to display the objects at the end of a pipeline. In
+> PowerShell, it is generally not necessary to use the cmdlet in instances where the output is
+> displayed by default. For example, `Get-Process | Write-Output` is equivalent to `Get-Process`.
+> Or, `echo "Home directory: $HOME"` can be written, `"Home directory: $HOME"`.
```
Comment (mklement0) [inline reference/5.1/Microsoft.PowerShell.Utility/Write-Output.md:36]: I like it, but I still think that this is the place to _briefly_ note the one use where explicit use of `Write-Output` _is_ useful: `Write-Output -NoEnumerate`, to output a collection _as a single object_, so that the next downstream command receives it as such.

You could just point to the `-NoEumerate` parameter description for details. Incidentally, I think the current description doesn't do a great job of clarifying the parameter's purpose, and "The NoEnumerate parameter has no effect on collections that were created by wrapping commands in parentheses" is downright confusing. What matters is that if you wrap the `Write-Output` command _as a whole_ in `(...)`, the effects of `-NoEnumerate` are negated.

Also - sorry to pack things on here - I don't think the examples should use `@(...)` so as to perpetuate the misconception that `@(...)` is needed for array literals.

Comment (chasewilson) [reply reference/5.1/Microsoft.PowerShell.Utility/Write-Output.md:36]: No need to be sorry, this is all great input. 

1. To the first point, I went ahead and added it to the descriptiont. However, I took out the NOTE identifier as I think adding all of this info will read better as just a full description. Let me know what you think about the explaination and if you have a reason for keeping it in a NOTE format. 
2. I agree with you here and I went ahead and updated to show the actual behavior. I think it's clear but feel free to let me know what you think.
3. Not at all. This is true and I struggle with using the operator unnecessarily. I've removed them from the examples

Comment (mklement0) [reply reference/5.1/Microsoft.PowerShell.Utility/Write-Output.md]: All looks good now - thank you.

**Comment by chasewilson** [comment]
#sign-off

File: reference/5.1/Microsoft.PowerShell.Utility/Write-Output.md [hunk 2]
```diff
@@ -28,10 +28,16 @@ If the command is the last command in the pipeline, the object is displayed in t
...
-to `Get-Process`.
+This cmdlet is typically used in scripts to display strings and other objects on the console. One of
+the built-in aliases for `Write-Output` is `echo` and similar to other shells that use `echo`, the
+default behavior is to display the output at the end of a pipeline. In PowerShell, it is generally
+not necessary to use the cmdlet in instances where the output is displayed by default. For example,
+`Get-Process | Write-Output` is equivalent to `Get-Process`. Or, `echo "Home directory: $HOME"` can
+be written, `"Home directory: $HOME"`.
+
+By default, `Write-Output` enumerates through collections provided to the cmdlet. However,
+`Write-Host` can also be used to pass collections down the pipeline as a single object with the
```
Comment (DCtheGeek) [inline reference/5.1/Microsoft.PowerShell.Utility/Write-Output.md:39]: Should this Write-Host be Write-Output?

File: reference/5.1/Microsoft.PowerShell.Utility/Write-Output.md [hunk 3]
```diff
@@ -101,14 +107,14 @@ Accept wildcard characters: False
 
 By default, the `Write-Output` cmdlet always enumerates its output. The **NoEnumerate** parameter
 suppresses the default behavior, and prevents `Write-Output` from enumerating output. The
-**NoEnumerate** parameter has no effect on collections that were created by wrapping commands in
-parentheses, because the parentheses force enumeration.
+**NoEnumerate** parameter has no effect if the command is wrapped in parentheses, because the
+parentheses force enumeration. For example, `(Write-Output 1,2,3)` will still enumerate the array.
```
Comment (DCtheGeek) [inline reference/5.1/Microsoft.PowerShell.Utility/Write-Output.md:111]: Remove the passive voice with "will".

File: reference/5.1/Microsoft.PowerShell.Utility/Write-Output.md [hunk 4]
```diff
@@ -101,14 +107,14 @@ Accept wildcard characters: False
...
 
 > [!IMPORTANT]
-> There is an issue with this switch in Windows PowerShell that is fixed in PowerShell 6.2 and above.
-> When using **NoEnumerate** and explicitly using the **InputObject** parameter, the command will
-> still enumerate. The workaround is to positionally define the **InputObject** parameter and the
-> enumeration will be suppressed.
+> There is an issue with this switch in Windows PowerShell that is fixed in PowerShell 6.2 and
+> above. When using **NoEnumerate** and explicitly using the **InputObject** parameter, the command
+> will still enumerate. The workaround is to positionally define the **InputObject** parameter and
```
Comment (DCtheGeek) [inline reference/5.1/Microsoft.PowerShell.Utility/Write-Output.md:116]: I know this existed before, but fix the passive voice here, too.

**Review by DCtheGeek — CHANGES_REQUESTED** [review]

File: reference/5.1/Microsoft.PowerShell.Utility/Write-Output.md (hunk 2 above)
Comment (mklement0) [reply reference/5.1/Microsoft.PowerShell.Utility/Write-Output.md:39]: Good catch: indeed it should be `Write-Output`.

File: reference/5.1/Microsoft.PowerShell.Utility/Write-Output.md (hunk 4 above)
Comment (mklement0) [reply reference/5.1/Microsoft.PowerShell.Utility/Write-Output.md:116]: While we're at it: "The workaround is to positionally define the **InputObject** parameter" is confusing.

I suggest:
"The workaround is to pass the arguments positionally, without naming `-InputObject`." 

Rationale: no _parameter_ is being _defined_; parameter _values_ or _arguments_ are being _passed_. 

A brief tangent: it's worth getting clarity on the language around these concepts throughout all topics:
* _parameter_ is the value _holder_ that is part of the command's definition, which is _bound_ during argument passing on invocation.
* what is bound to it is either a _parameter value_ or, perhaps better, an _argument_.

Arguments can be _positional_ or _named_ (preceded by their target parameter _name_)

Comment (chasewilson) [reply reference/5.1/Microsoft.PowerShell.Utility/Write-Output.md:116]: I would argue that, that wording makes it more confusing. You are correct that the Parameter is the value holder and in effect is what the function or cmdlet will interact with.  Even positionally defining a parameter, the parameter is what is referenced when used again and not the argument. The parameter is implicitly defined, not "no parameter value is being defined."

File: reference/5.1/Microsoft.PowerShell.Utility/Write-Output.md [hunk 5]
```diff
@@ -108,13 +108,13 @@ Accept wildcard characters: False
...
 **NoEnumerate** parameter has no effect if the command is wrapped in parentheses, because the
-parentheses force enumeration. For example, `(Write-Output 1,2,3)` will still enumerate the array.
+parentheses force enumeration. For example, `(Write-Output 1,2,3)` still enumerates the array.
 
 > [!IMPORTANT]
 > There is an issue with this switch in Windows PowerShell that is fixed in PowerShell 6.2 and
 > above. When using **NoEnumerate** and explicitly using the **InputObject** parameter, the command
-> will still enumerate. The workaround is to positionally define the **InputObject** parameter and
-> the enumeration will be suppressed.
+> still enumerates. To workaround this, positionally defining the **InputObject** parameter
```
[the remark this answers is not in the document]
Comment (chasewilson) [reply reference/5.1/Microsoft.PowerShell.Utility/Write-Output.md:116]: How about a compromise? 
"To work around this, pass the **InputObject** argument positionally."

[the remark this answers is not in the document]
Comment (mklement0) [reply reference/5.1/Microsoft.PowerShell.Utility/Write-Output.md:116]: Well, I can't just give in without another concession: how about "argument(s)"? 😜

[the remark this answers is not in the document]
Comment (mklement0) [reply reference/5.1/Microsoft.PowerShell.Utility/Write-Output.md:116]: Kidding aside: 👍 (though note that you can do `Write-Output 1 2` - individually passed arguments (ValueFromRemainingArguments) - in lieu of `Write-Output -InputObject 1, 2`)

[the remark this answers is not in the document]
Comment (chasewilson) [reply reference/5.1/Microsoft.PowerShell.Utility/Write-Output.md:116]: 😄 Think I can live with that concession. You have a deal 🤝

**Review by DCtheGeek — APPROVED** [review]

File: reference/5.1/Microsoft.PowerShell.Utility/Write-Output.md (hunk 4 above)
Comment (mklement0) [reply reference/5.1/Microsoft.PowerShell.Utility/Write-Output.md]: @chasewilson , there seems to be problem with GitHub's email notifications - I only just saw this now, after our later conversation.

To be clear: I meant that it's important to distinguish between a _parameter_ - which is _defined_, as part of the command's _implementation_  - and its _value_ (its _argument_), which is _passed_ (or _specified_), as part of the command's _invocation_, and then _bound_ to the parameter.

Referring to both the parameter itself and a value situationally bound to it as _parameter_ invites confusion.
Similarly, I  suggest reserving _defining_ for implementations aspect vs. _passing_ or _specifying_ for the invocation aspect.

To put it differently: you're passing _arguments_ (parameter _values_) to a _parameter_, either _positionally_ or _named_ (by explicitly specifying the parameter's name).
